// rbx-reader/dist/AttributesParser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parse = void 0;
// utils
class ByteReader {
    constructor(buf) {
        this.cursor = 0;
        this.buf = buf;
    }
    readUInt32LE() {
        const value = this.buf.readUInt32LE(this.cursor);
        this.cursor += 4;
        return value;
    }
    readUInt16LE() {
        const value = this.buf.readUInt16LE(this.cursor);
        this.cursor += 2;
        return value;
    }
    readUInt8() {
        const value = this.buf.readUInt8(this.cursor);
        this.cursor += 1;
        return value;
    }
    readInt32LE() {
        const value = this.buf.readInt32LE(this.cursor);
        this.cursor += 4;
        return value;
    }
    readFloatLE() {
        const value = this.buf.readFloatLE(this.cursor);
        this.cursor += 4;
        return value;
    }
    readDoubleLE() {
        const value = this.buf.readDoubleLE(this.cursor);
        this.cursor += 8;
        return value;
    }
    readString() {
        const length = this.readUInt32LE();
        const value = this.buf.subarray(this.cursor, this.cursor + length).toString('utf8');
        this.cursor += length;
        return value;
    }
}
function parse(buf) {
    const reader = new ByteReader(buf);
    const LENGTH = reader.readUInt32LE();
    console.log(`we're expecting ${LENGTH} attr's`);
    const attrs = new Map();
    for (let _ = 0; _ < LENGTH; _++) {
        const startCursor = reader.cursor;
        const key = reader.readString();
        console.log(`reading key ${key} from ${startCursor}`);
        const type = getTypeFromNumber(reader.readUInt8());
        console.log(`key ${key} is type ${type}`);
        switch (type) {
            case Type.BinaryString: {
                const str = reader.readString();
                attrs.set(key, str);
                break;
            }
            case Type.Bool: {
                attrs.set(key, reader.readUInt8() != 0);
                break;
            }
            case Type.BrickColor: {
                const value = reader.readUInt32LE();
                attrs.set(key, { BrickColor: value });
                break;
            }
            case Type.CFrame: {
                const vector = new Vector3(reader.readFloatLE(), reader.readFloatLE(), reader.readFloatLE());
                const rotId = reader.readUInt8();
                let rotation;
                if (rotId == 0) {
                    const thing = [];
                    for (let __ = 0; __ < 3; __++) {
                        thing.push(new Vector3(reader.readFloatLE(), reader.readFloatLE(), reader.readFloatLE()));
                    }
                    rotation = new Matrix3(thing[0], thing[1], thing[2]);
                }
                else {
                    rotation = Matrix3.fromRotationId(rotId);
                }
                if (!rotation)
                    throw 'missing rotation matrix';
                else
                    console.log(rotation);
                attrs.set(key, new CFrame(vector, rotation));
                break;
            }
            case Type.Color3: {
                attrs.set(key, new Color3(reader.readFloatLE(), reader.readFloatLE(), reader.readFloatLE()));
                break;
            }
            case Type.ColorSequence: {
                const keypointCount = reader.readUInt32LE();
                const sequence = new ColorSequence([]);
                for (let kp = 0; kp < keypointCount; kp++) {
                    sequence.keypoints.push(new ColorSequenceKeypoint(reader.readFloatLE(), reader.readFloatLE(), new Color3(reader.readFloatLE(), reader.readFloatLE(), reader.readFloatLE())));
                }
                attrs.set(key, sequence);
                break;
            }
            case Type.Float32: {
                attrs.set(key, reader.readFloatLE());
                break;
            }
            case Type.Float64: {
                attrs.set(key, reader.readDoubleLE());
                break;
            }
            case Type.Font: {
                const weight = reader.readUInt16LE();
                const style = reader.readUInt8();
                const family = reader.readString();
                const cfId = reader.readString();
                attrs.set(key, new Font(weight, style, family, cfId));
                break;
            }
            case Type.NumberRange: {
                attrs.set(key, new NumberRange(reader.readFloatLE(), reader.readFloatLE()));
                break;
            }
            case Type.NumberSequence: {
                const keypointCount = reader.readUInt32LE();
                const sequence = new NumberSequence([]);
                for (let kp = 0; kp < keypointCount; kp++) {
                    sequence.keypoints.push(new NumberSequenceKeypoint(reader.readFloatLE(), reader.readFloatLE(), reader.readFloatLE()));
                }
                attrs.set(key, sequence);
                break;
            }
            case Type.Rect: {
                attrs.set(key, new Rect(new Vector2(reader.readFloatLE(), reader.readFloatLE()), new Vector2(reader.readFloatLE(), reader.readFloatLE())));
                break;
            }
            case Type.UDim: {
                attrs.set(key, new UDim(reader.readFloatLE(), reader.readInt32LE()));
                break;
            }
            case Type.UDim2: {
                attrs.set(key, new UDim2(new UDim(reader.readFloatLE(), reader.readInt32LE()), new UDim(reader.readFloatLE(), reader.readInt32LE())));
                break;
            }
            case Type.Vector2: {
                attrs.set(key, new Vector2(reader.readFloatLE(), reader.readFloatLE()));
                break;
            }
            case Type.Vector3: {
                attrs.set(key, new Vector3(reader.readFloatLE(), reader.readFloatLE(), reader.readFloatLE()));
                break;
            }
        }
        const endCursor = reader.cursor;
        console.log(`read ${key} (${type}) from ${startCursor} to ${endCursor}`);
        console.log(`subarray: ${reader.buf.subarray(startCursor, endCursor).toString('hex')}`);
        console.log(`${key} = ${JSON.stringify(attrs.get(key))} [END]`);
    }
}
exports.parse = parse;
var Type;
(function (Type) {
    Type[Type["BinaryString"] = 2] = "BinaryString";
    Type[Type["Bool"] = 3] = "Bool";
    Type[Type["Float32"] = 5] = "Float32";
    Type[Type["Float64"] = 6] = "Float64";
    Type[Type["UDim"] = 9] = "UDim";
    Type[Type["UDim2"] = 10] = "UDim2";
    Type[Type["BrickColor"] = 14] = "BrickColor";
    Type[Type["Color3"] = 15] = "Color3";
    Type[Type["Vector2"] = 16] = "Vector2";
    Type[Type["Vector3"] = 17] = "Vector3";
    Type[Type["CFrame"] = 20] = "CFrame";
    Type[Type["NumberSequence"] = 23] = "NumberSequence";
    Type[Type["ColorSequence"] = 25] = "ColorSequence";
    Type[Type["NumberRange"] = 27] = "NumberRange";
    Type[Type["Rect"] = 28] = "Rect";
    Type[Type["Font"] = 33] = "Font";
})(Type || (Type = {}));
const types = {
    0x02: Type.BinaryString,
    0x03: Type.Bool,
    0x05: Type.Float32,
    0x06: Type.Float64,
    0x09: Type.UDim,
    0x0A: Type.UDim2,
    0x0E: Type.BrickColor,
    0x0F: Type.Color3,
    0x10: Type.Vector2,
    0x11: Type.Vector3,
    0x14: Type.CFrame,
    0x17: Type.NumberSequence,
    0x19: Type.ColorSequence,
    0x1B: Type.NumberRange,
    0x1C: Type.Rect,
    0x21: Type.Font
};
function getTypeFromNumber(num) {
    return types[num];
}
class Vector2 {
    constructor(x = 0, y = 0) {
        this.x = 0;
        this.y = 0;
        this.x = x;
        this.y = y;
    }
}
class Vector3 {
    constructor(x = 0, y = 0, z = 0) {
        this.x = 0;
        this.y = 0;
        this.z = 0;
        this.x = x;
        this.y = y;
        this.z = z;
    }
}
class Matrix3 {
    constructor(x, y, z) {
        this.x = x;
        this.y = y;
        this.z = z;
    }
    static fromRotationId(rotId) {
        return Matrix3.ROTATION_MATRICES[rotId];
    }
}
Matrix3.identity = new Matrix3(new Vector3(1, 0, 0), new Vector3(0, 1, 0), new Vector3(0, 0, 1));
Matrix3.ROTATION_MATRICES = {
    0x02: Matrix3.identity,
    0x03: new Matrix3(new Vector3(1, 0, 0), new Vector3(0, 0, -1), new Vector3(0, 1, 0)),
    0x05: new Matrix3(new Vector3(1, 0, 0), new Vector3(0, -1, 0), new Vector3(0, 0, -1)),
    0x06: new Matrix3(new Vector3(1, 0, 0), new Vector3(0, 0, 1), new Vector3(0, -1, 0)),
    0x07: new Matrix3(new Vector3(0, 1, 0), new Vector3(1, 0, 0), new Vector3(0, 0, -1)),
    0x09: new Matrix3(new Vector3(0, 0, 1), new Vector3(1, 0, 0), new Vector3(0, 1, 1)),
    0x0a: new Matrix3(new Vector3(0, -1, 0), new Vector3(1, 0, 0), new Vector3(0, 0, 1)),
    0x0c: new Matrix3(new Vector3(0, 0, -1), new Vector3(1, 0, 0), new Vector3(0, -1, 0)),
    0x0d: new Matrix3(new Vector3(0, 1, 0), new Vector3(0, 0, 1), new Vector3(1, 0, 0)),
    0x0e: new Matrix3(new Vector3(0, 0, -1), new Vector3(0, 1, 0), new Vector3(1, 0, 1)),
    0x10: new Matrix3(new Vector3(0, -1, 0), new Vector3(0, 0, -1), new Vector3(1, 0, 0)),
    0x11: new Matrix3(new Vector3(0, 0, 1), new Vector3(0, -1, 0), new Vector3(1, 0, 0)),
    0x14: new Matrix3(new Vector3(-1, 0, 0), new Vector3(0, 1, 0), new Vector3(0, 0, -1)),
    0x15: new Matrix3(new Vector3(-1, 0, 0), new Vector3(0, 0, 1), new Vector3(0, 1, 0)),
    0x17: new Matrix3(new Vector3(-1, 0, 0), new Vector3(0, -1, 0), new Vector3(0, 0, 1)),
    0x18: new Matrix3(new Vector3(-1, 0, 0), new Vector3(0, 0, -1), new Vector3(0, -1, 0)),
    0x19: new Matrix3(new Vector3(0, 1, 0), new Vector3(-1, 0, 0), new Vector3(0, 0, 1)),
    0x1b: new Matrix3(new Vector3(0, 0, -1), new Vector3(-1, 0, -1), new Vector3(0, 1, 0)),
    0x1c: new Matrix3(new Vector3(0, -1, 0), new Vector3(-1, 0, 0), new Vector3(0, 0, -1)),
    0x1e: new Matrix3(new Vector3(0, 0, 1), new Vector3(-1, 0, 0), new Vector3(0, -1, 1)),
    0x1f: new Matrix3(new Vector3(0, 1, 0), new Vector3(0, 0, -1), new Vector3(-1, 0, 0)),
    0x20: new Matrix3(new Vector3(0, 0, 1), new Vector3(0, 1, 0), new Vector3(-1, 0, 0)),
    0x22: new Matrix3(new Vector3(0, -1, 0), new Vector3(0, 0, 1), new Vector3(-1, 0, 0)),
    0x23: new Matrix3(new Vector3(0, 0, -1), new Vector3(0, -1, 0), new Vector3(-1, 0, 0)),
};
class CFrame {
    constructor(position, orientation) {
        this.position = position;
        this.orientation = orientation;
    }
}
class Color3 {
    get r() {
        return this._r;
    }
    set r(r) {
        this._r = Math.max(0, Math.min(r, 1));
    }
    get g() {
        return this._g;
    }
    set g(g) {
        this._g = Math.max(0, Math.min(g, 1));
    }
    get b() {
        return this._b;
    }
    set b(b) {
        this._b = Math.max(0, Math.min(b, 1));
    }
    /**
     * @param r red [0,1]
     * @param g green [0,1]
     * @param b blue [0,1]
     */
    constructor(r = 0, g = 0, b = 0) {
        this._r = 0;
        this._g = 0;
        this._b = 0;
        this._r = r;
        this._g = g;
        this._b = b;
    }
}
class Font {
    constructor(weight, style, family, cfid) {
        this.weight = weight;
        this.style = style;
        this.family = family;
        this.cachedFaceId = cfid;
    }
}
class NumberRange {
    get Min() { return this.min; }
    get Max() { return this.max; }
    numberInRange(num, ignoreBottom = false, ignoreTop = false) {
        return (ignoreBottom ? num > this.min : num >= this.min) && (ignoreTop ? num < this.max : num <= this.max);
    }
    constructor(min, max) {
        this.min = min;
        this.max = max;
    }
}
class NumberSequenceKeypoint {
    constructor(envelope, time, value) {
        this.envelope = envelope;
        this.time = time;
        this.value = value;
    }
}
class NumberSequence {
    constructor(keypoints) {
        this.keypoints = [];
        this.keypoints.push(...keypoints);
    }
}
class ColorSequenceKeypoint {
    constructor(envelope, time, value) {
        this.envelope = 0;
        this.envelope = envelope;
        this.time = time;
        this.value = value;
    }
}
class ColorSequence {
    constructor(keypoints) {
        this.keypoints = [];
        this.keypoints.push(...keypoints);
    }
}
class Rect {
    constructor(min, max) {
        this.min = min;
        this.max = max;
    }
}
class UDim {
    constructor(scale, offset) {
        this.scale = 0;
        this.offset = 0;
        this.scale = scale;
        this.offset = offset;
    }
}
class UDim2 {
    constructor(x, y) {
        this.x = x;
        this.y = y;
    }
}
