// rbx-reader/dist/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseAttributes = exports.readFile = exports.parseBuffer = void 0;
const ByteReader_1 = __importDefault(require("./ByteReader"));
const BinaryParser_1 = __importDefault(require("./BinaryParser"));
const XmlParser_1 = __importDefault(require("./XmlParser"));
const node_fs_1 = require("node:fs");
const attributes_parser_1 = require("./attributes-parser/attributes_parser");
const Instance_1 = require("./Instance");
function parseBuffer(buffer) {
    try {
        return XmlParser_1.default.parse(buffer);
    }
    catch { }
    const reader = new ByteReader_1.default(buffer);
    const header = reader.String(7);
    if (header !== '<roblox')
        throw new Error('Invalid RBXM/RBXL file');
    if (reader.Byte() === 0x21) {
        const { result, instances } = BinaryParser_1.default.parse(buffer);
        return { result, instances };
    }
    throw new Error('Invalid file');
}
exports.parseBuffer = parseBuffer;
function readFile(file) {
    return parseBuffer((0, node_fs_1.readFileSync)(file));
}
exports.readFile = readFile;
function parseAttributes(attributesSerializeBuf) {
    const result = (0, attributes_parser_1.parse_attrs)(Buffer.from(attributesSerializeBuf));
    const Attributes = {};
    result.forEach((v, k) => {
        Attributes[k] = v;
    });
    const finalAttrs = {};
    Object.keys(Attributes).forEach(k => {
        const v = Attributes[k];
        if (Instance_1.AttrAssertions.isBoolAttr(v))
            finalAttrs[k] = v.Bool;
        else if (Instance_1.AttrAssertions.isStringAttr(v))
            finalAttrs[k] = v.BinaryString;
        else if (Instance_1.AttrAssertions.isNumberSequenceAttr(v))
            finalAttrs[k] = v.NumberSequence;
        else if (Instance_1.AttrAssertions.isRectAttr(v))
            finalAttrs[k] = v.Rect;
        else if (Instance_1.AttrAssertions.isCFrameAttr(v))
            finalAttrs[k] = v.CFrame;
        else if (Instance_1.AttrAssertions.isNumberRangeAttr(v))
            finalAttrs[k] = {
                lower: v.NumberRange[0],
                upper: v.NumberRange[1]
            };
        else if (Instance_1.AttrAssertions.isColorSequenceAttr(v))
            finalAttrs[k] = v.ColorSequence;
        else if (Instance_1.AttrAssertions.isVector3Attr(v))
            finalAttrs[k] = {
                x: v.Vector3[0],
                y: v.Vector3[1],
                z: v.Vector3[2]
            };
        else if (Instance_1.AttrAssertions.isUDim2Attr(v))
            finalAttrs[k] = {
                x: {
                    scale: v.UDim2[0][0],
                    offset: v.UDim2[0][1]
                },
                y: {
                    scale: v.UDim2[1][0],
                    offset: v.UDim2[1][1]
                }
            };
        else if (Instance_1.AttrAssertions.isVector2Attr(v))
            finalAttrs[k] = {
                x: v.Vector2[0],
                y: v.Vector2[1]
            };
        else if (Instance_1.AttrAssertions.isColor3Attr(v))
            finalAttrs[k] = {
                r: v.Color3[0],
                g: v.Color3[1],
                b: v.Color3[2]
            };
        else if (Instance_1.AttrAssertions.isNumberAttr(v))
            finalAttrs[k] = v.Float64;
        else if (Instance_1.AttrAssertions.isUDimAttr(v))
            finalAttrs[k] = {
                scale: v.UDim[0],
                offset: v.UDim[1]
            };
        else if (Instance_1.AttrAssertions.isBrickColorAttr(v))
            finalAttrs[k] = {
                ...v
            };
    });
    return Attributes;
}
exports.parseAttributes = parseAttributes;
