// rbx-reader/dist/Instance.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.InstanceRoot = exports.Instance = exports.AttrAssertions = void 0;
const assert_1 = __importDefault(require("assert"));
const InstanceUtils = {
    findFirstChild(target, name, recursive = false) {
        const children = target instanceof Instance ? target.Children : target;
        for (const child of children) {
            if (child.getProperty('Name') === name) {
                return child;
            }
        }
        if (recursive) {
            const arrays = [children];
            while (arrays.length) {
                for (const desc of arrays.shift()) {
                    if (desc.getProperty('Name') === name) {
                        return desc;
                    }
                    if (desc.Children.length) {
                        arrays.push(desc.Children);
                    }
                }
            }
        }
        return undefined;
    },
    findFirstChildOfClass(target, className, recursive = false) {
        const children = target instanceof Instance ? target.Children : target;
        for (const child of children) {
            if (child.getProperty('ClassName') === className) {
                return child;
            }
        }
        if (recursive) {
            const arrays = [children];
            while (arrays.length > 0) {
                for (const desc of arrays.shift()) {
                    if (desc.getProperty('ClassName') === className) {
                        return desc;
                    }
                    if (desc.Children.length) {
                        arrays.push(desc.Children);
                    }
                }
            }
        }
        return undefined;
    }
};
class InstanceRoot extends Array {
    getChildren() {
        return this;
    }
    getDescendants() {
        const l = this.getChildren();
        l.forEach(v => l.push(...v.getDescendants()));
        return l;
    }
    findFirstChild(name, recursive = false) { return InstanceUtils.findFirstChild(this, name, recursive); }
    findFirstChildOfClass(className, recursive = false) { return InstanceUtils.findFirstChildOfClass(this, className, recursive); }
}
exports.InstanceRoot = InstanceRoot;
function isBoolAttr(obj) {
    return typeof obj == 'object' && typeof obj.Bool == 'boolean';
}
function isStringAttr(obj) {
    return typeof obj == 'object' && typeof obj.BinaryString == 'string';
}
function isNumSequenceKeyPtAttr(obj) {
    return typeof obj == 'object' && typeof obj.time == 'number' && typeof obj.value == 'number' && typeof obj.envelope == 'number';
}
function isNumberSequenceAttr(obj) {
    return typeof obj == 'object' && typeof obj.NumberSequence == 'object' &&
        Array.isArray(obj.NumberSequence.keypoints) && obj.NumberSequence.keypoints.every(isNumSequenceKeyPtAttr);
}
function isRectAttr(obj) {
    return typeof obj == 'object' && Array.isArray(obj.Rect) && obj.Rect.length == 2 &&
        obj.Rect.every((v) => Array.isArray(v) && v.length == 2);
}
function isCFrameAttr(obj) {
    return typeof obj == 'object' && typeof obj.CFrame == 'object' && Array.isArray(obj.CFrame.position) &&
        obj.CFrame.position.length == 3 && obj.CFrame.position.every((v) => typeof v == 'number') &&
        Array.isArray(obj.CFrame.orientation) && obj.CFrame.orientation.length == 3 &&
        obj.CFrame.orientation.every((v) => Array.isArray(v) &&
            v.length == 3 && v.every((v2) => typeof v2 == 'number'));
}
function isNumberRangeAttr(obj) {
    return typeof obj == 'object' && Array.isArray(obj.NumberRange) &&
        obj.NumberRange.every((v) => typeof v == 'number') && obj.NumberRange.length == 2;
}
function isColorSequenceKeyPtAttr(obj) {
    return typeof obj == 'object' && typeof obj.time == 'number' && Array.isArray(obj.color)
        && obj.color.every((v) => typeof v == 'number') && obj.color.length == 3;
}
function isColorSequenceAttr(obj) {
    return typeof obj == 'object' && typeof obj.ColorSequence == 'object' &&
        Array.isArray(obj.ColorSequence.keypoints) && obj.ColorSequence.keypoints.every(isColorSequenceKeyPtAttr);
}
function isVector3Attr(obj) {
    return typeof obj == 'object' && Array.isArray(obj.Vector3) &&
        obj.Vector3.every((v) => typeof v == 'number') && obj.Vector3.length == 3;
}
function isUDim2Attr(obj) {
    return typeof obj == 'object' && Array.isArray(obj.UDim2) && obj.UDim2.length == 2 &&
        obj.UDim2.every((v) => Array.isArray(v) && v.every((v2) => typeof v2 == 'number') &&
            v.length == 2);
}
function isVector2Attr(obj) {
    return typeof obj == 'object' && Array.isArray(obj.Vector2) &&
        obj.Vector2.every((v) => typeof v == 'number') && obj.Vector2.length == 2;
}
function isColor3Attr(obj) {
    return typeof obj == 'object' && Array.isArray(obj.Color3) &&
        obj.Color3.every((v) => typeof v == 'number') && obj.Color3.length == 3;
}
function isBrickColorAttr(obj) {
    return typeof obj == 'object' && typeof obj.BrickColor == 'number';
}
function isNumberAttr(obj) {
    return typeof obj == 'object' && typeof obj.Float64 == 'number';
}
function isUDimAttr(obj) {
    return typeof obj == 'object' && Array.isArray(obj.UDim) && obj.UDim.length == 2 &&
        obj.UDim.every((v) => typeof v == 'number');
}
exports.AttrAssertions = {
    isBoolAttr,
    isStringAttr,
    isNumberSequenceAttr,
    isRectAttr,
    isCFrameAttr,
    isNumberRangeAttr,
    isColorSequenceAttr,
    isVector3Attr,
    isUDim2Attr,
    isVector2Attr,
    isColor3Attr,
    isBrickColorAttr,
    isNumberAttr,
    isUDimAttr,
    isFloat64Attr: isNumberAttr
};
class Instance {
    get ClassName() {
        return this.Properties['ClassName'].value;
    }
    get Name() {
        return this.Properties['Name'].value;
    }
    get Parent() {
        return this.Properties['Parent'].value;
    }
    getFullName() {
        if (this.Parent && !(this.Parent instanceof InstanceRoot)) {
            return `${this.Parent.getFullName()}.${this.Name}`;
        }
        else {
            return `${this.Name}`;
        }
    }
    static new(className) {
        (0, assert_1.default)(typeof className === 'string', 'className is not a string');
        return new Instance(className);
    }
    constructor(className) {
        this.Children = [];
        this.Properties = {};
        this.Attributes = {};
        (0, assert_1.default)(typeof className === 'string', 'className is not a string');
        this.Children = [];
        this.setProperty('ClassName', className, 'string');
        this.setProperty('Name', 'Instance', 'string');
        this.setProperty('Parent', undefined, 'Instance');
    }
    setProperty(name, value, type) {
        if (!type) {
            if (typeof value === 'boolean') {
                type = 'bool';
            }
            else if (value instanceof Instance) {
                type = 'Instance';
            }
            else {
                throw new TypeError('You need to specify property type since it cannot be inferred');
            }
        }
        var descriptor = this.Properties[name];
        if (descriptor) {
            (0, assert_1.default)(descriptor.type === type, `Property type mismatch: ${type} !== ${descriptor.type}`);
            if (name === 'Parent' && descriptor.value instanceof Instance) {
                const index = descriptor.value.Children.indexOf(this);
                if (index !== -1)
                    descriptor.value.Children.splice(index, 1);
            }
            descriptor.value = value;
        }
        else
            descriptor = this.Properties[name] = { type, value };
        if (name === 'Parent' && descriptor.value instanceof Instance)
            descriptor.value.Children.push(this);
        if (name !== 'Children' && name !== 'Properties' && !(name in Object.getPrototypeOf(this)))
            this[name] = value;
    }
    getProperty(name, caseInsensitive = false) {
        const descriptor = this.Properties[name] || caseInsensitive && Object.entries(this.Properties).find(x => x[0].toLowerCase() === name.toLowerCase())?.[1];
        return descriptor ? descriptor.value : undefined;
    }
    getChildren() {
        return this.Children;
    }
    getDescendants() {
        const desc = this.Children;
        desc.forEach(inst => {
            desc.push(...inst.getDescendants());
        });
        return desc;
    }
    hasProperty(name, caseInsensitive = false) {
        return name in this.Properties || caseInsensitive && !Object.entries(this.Properties).find(x => x[0].toLowerCase() === name.toLowerCase());
    }
    findFirstChild(name, recursive = false) { return InstanceUtils.findFirstChild(this, name, recursive); }
    findFirstChildOfClass(className, recursive = false) { return InstanceUtils.findFirstChildOfClass(this, className, recursive); }
    getAttribute(name) {
        return this.Attributes[name];
    }
}
exports.Instance = Instance;
